luarules/gadgets/game_unit_sharing_mode.lua

```diff
@@ -0,0 +1,101 @@
+local gadget = gadget ---@type Gadget
+
+function gadget:GetInfo()
+	return {
+		name    = 'Unit Sharing Control',
+		desc    = 'Controls unit sharing based on modoption settings',
+		author  = 'Rimilel',
+		date    = 'May 2024 / April 2025',
+		license = 'GNU GPL, v2 or later',
+		layer   = 0,
+		enabled = true
+	}
+end
+
+----------------------------------------------------------------
+-- Synced only
+----------------------------------------------------------------
+if not gadgetHandler:IsSyncedCode() then
+	return false
+end
+
+local unitSharingMode = Spring.GetModOptions().unit_sharing_mode or "enabled"
+local unitMarketEnabled = Spring.GetModOptions().unit_market or false
+
+-- Disable the gadget only if unit sharing is fully enabled
+if unitSharingMode == "enabled" then
+	return false
+end
+
+--[[ Engine Callin Note: AllowUnitTransfer blocks if *any* gadget returns false.
+     This gadget must therefore explicitly check for and allow certain transfer types
+     (Market, /take, Capture) to prevent this gadget's rules (disabled/t2cons)
+     from blocking valid engine or other gadget actions.
+]]
+
+-- Flags for specific transfer types that should cleanly bypass this gadget's sharing rules.
+local transferOverrideFlags = {
+	"marketTransferInProgress",
+	-- Note: takeCommand logic uses a separate check function due to engine limitations.
+}
+
+local function isT2Constructor(unitDef)
+	if not unitDef then return false end
+
+	return not unitDef.isFactory
+			and #(unitDef.buildOptions or {}) > 0
+			and unitDef.customParams.techlevel == "2"
+end
+
+-- Handles the specific condition for allowing /take transfers
+-- Returns true if the transfer should be allowed due to being a /take, false otherwise.
+local function CheckTakeCondition(fromTeamID, toTeamID)
+	-- Check if sender is allied and has no active human players
+	if Spring.AreTeamsAllied(fromTeamID, toTeamID) then
+		local senderHasActivePlayer = false
+		for _, playerID in ipairs(Spring.GetPlayerList()) do
+			local _, active, spectator, teamID = Spring.GetPlayerInfo(playerID)
+			if active and not spectator and teamID == fromTeamID then
+				senderHasActivePlayer = true
```

Comment:
You can return true directly here right, no need for the variable (and check later)